data-import (#99)
cienciadedatos/pt-r4ds: data-import.qmd

# Importação de dados {#sec-data-import}

```{r}
#| echo: false

source("_common.R")
```

## Introdução

Trabalhar com dados fornecidos por pacotes R é uma ótima maneira de aprender ferramentas de ciência de dados, mas você irá querer aplicar o que aprendeu aos seus próprios dados em algum momento.

> Neste capítulo, você aprenderá o básico de como ler arquivos de dados no R.

Reviewer: ```suggestion
Neste capítulo, você aprenderá o básico sobre como ler arquivos de dados no R.
```

Especificamente, esse capítulo focará na leitura de arquivos de texto retangulares.
Começaremos com conselhos práticos para lidar com recursos como nomes de colunas, tipos e dados ausentes.

> Você então aprenderá sobre como ler dados de múltiplos arquivos de uma vez e exportar dados do R para um arquivo.

Reviewer: ```suggestion
Você então aprenderá sobre como ler dados de múltiplos arquivos de uma única vez e exportar dados do R para um arquivo.
```

Finalmente, você aprenderá a criar manualmente data frames no R.

### Pré-requisitos

> Neste capítulo, você irá aprender como importar arquivos planos no R com o pacote **readr**, que faz parte do tidyverse.

Reviewer: Eu acho que arquivos de texto ainda não é exatamente a ideia. Talvez arquivos flat sejam arquivos **simples**?

> Neste capítulo, você irá aprender como importar arquivos planos no R com o pacote **readr**, que faz parte do tidyverse.

Reviewer 2: Eu fiz a sugestão pensando mais no formato que o readr lê mesmo... ele lê apenas arquivos de texto!

```

O código acima irá funcionar se você tiver o arquivo `estudantes.csv` em uma pasta chamada `data` no seu projeto.
Você pode baixar o arquivo `estudantes.csv` de <https://raw.githubusercontent.com/cienciadedatos/pt-r4ds/traducao-pt-2ed/data/estudantes.csv> ou você pode importá-lo diretamente dessa URL com:

```{r}
#| eval: false

estudantes <- read_csv("https://raw.githubusercontent.com/cienciadedatos/pt-r4ds/traducao-pt-2ed/data/estudantes.csv")
```

Quando você executa o comando `read_csv()`, ele imprime uma mensagem informando o número de linhas e colunas de dados, o delimitador que foi usado e as especificações das colunas (nomes das colunas organizados pelo tipo de dados que a coluna contém).
Ele também imprime algumas informações sobre como recuperar a especificação completa da coluna e como silenciar essa mensagem.
Essa mensagem é uma parte integral do readr, e voltaremos a ela em @sec-col-types.

### Dicas práticas

Após importar os dados, o próximo passo geralmente envolve transformá-los de um jeito que facilite o trabalho no restante da sua análise.
Com isso em mente, vamos dar uma segunda olhada no conjunto de dados `estudantes`.

```{r}
estudantes
```

> Na coluna `comida.favorita`, há uma série de itens de comida, e o texto (*string*) `"N/A"`, que deveria ter sido um `NA` real que o R reconheceria como um dado faltante (*not available*).

Reviewer 2: ```suggestion
Na coluna `comida.favorita`, há uma série de itens de comida, e o texto (*string*) `"N/A"`, que deveria ter sido um `NA` real que o R reconheceria como um dado faltante (*not available*).
Isso é algo que podemos resolver usando o argumento `na`.
```

library(tidyverse)
```

> ## Importando dados de um arquivo

Reviewer: ```suggestion
## Lendo dados de um arquivo
```

Me parece que é feita uma distinção dos autores do original entre _import_ e _read_. Talvez seja interessante manter essa diferença no português também.

De início, vamos focar no tipo de arquivo de dados retangulares mais comum: CSV, que é a sigla em inglês para valores separados por vírgula (*comma-separated values*)[^data-import-1].
Aqui está como um arquivo CSV simples se parece.
A primeira linha, comumente chamada de linha de cabeçalho, fornece os nomes das colunas, e as seis linhas seguintes fornecem os dados.
As colunas são separadas, ou delimitadas, por vírgulas.

```{r}
#| echo: false
#| message: false
#| comment: ""

read_lines("data/estudantes.csv") |> cat(sep = "\n")
```

@tbl-estudantes-table mostra uma representação dos mesmos dados como uma tabela.

```{r}
#| label: tbl-estudantes-table
#| echo: false
#| message: false
#| tbl-cap: Dados do arquivo estudantes.csv como uma tabela.

read_csv("data/estudantes.csv") |>
  knitr::kable()
```

> Podemos importar esse arquivo para o R usando `read_csv()`.

Reviewer: ```suggestion
Podemos ler esse arquivo em R usando `read_csv()`.
```

Mesma ideia do comentário anterior sobre manter distinção entre _import_ e _read_.

Quando você executa o comando `read_csv()`, ele imprime uma mensagem informando o número de linhas e colunas de dados, o delimitador que foi usado e as especificações das colunas (nomes das colunas organizados pelo tipo de dados que a coluna contém).
Ele também imprime algumas informações sobre como recuperar a especificação completa da coluna e como silenciar essa mensagem.
Essa mensagem é uma parte integral do readr, e voltaremos a ela em @sec-col-types.

### Practical advice
### Dicas práticas

Once you read data in, the first step usually involves transforming it in some way to make it easier to work with in the rest of your analysis.
Let's take another look at the `students` data with that in mind.
Após importar os dados, o próximo passo geralmente envolve transformá-los de um jeito que facilite o trabalho no restante da sua análise.
Com isso em mente, vamos dar uma segunda olhada no conjunto de dados `estudantes`.

```{r}
students
estudantes
```

In the `favourite.food` column, there are a bunch of food items, and then the character string `N/A`, which should have been a real `NA` that R will recognize as "not available".

> This is something we can address using the `na` argument.

Reviewer: Algum motivo especial para essa linha não ter sido traduzida?

Reviewer 2: Verdade! Não tinha reparado. fiz uma sugestão com a tradução dela

> Por padrão, `read_csv()` reconhece apenas strings vazias (`""`) como `NA`s, mas, para esse conjunto de dados, também queremos que ele reconheça a *string* de caracteres `"N/A"`.

Reviewer: ```suggestion
Por padrão, `read_csv()` reconhece apenas strings não preenchidos (`""`) como `NA`s, mas, para esse conjunto de dados, também queremos que ele reconheça os caracteres `"N/A"`.
```
A _string_ ou O _string_?
Pensei nessa alternativa de escrita para não precisarmos definir gênero para essa palavra.

Reviewer 2: Geralmente eu ouço chamarem string no feminino (e acabo falando assim também)!

> Para se referir a essas variáveis, você precisa cercá-las com crases, `` ` ``:

Reviewer: ```suggestion
Para se referir a essas variáveis, você precisa usar crases, `` ` ``:
```

```{r}
#| message: false
students <- read_csv("data/students.csv", na = c("N/A", ""))
estudantes <- read_csv("data/estudantes.csv", na = c("N/A", ""))

students
estudantes
```

You might also notice that the `Student ID` and `Full Name` columns are surrounded by backticks.
That's because they contain spaces, breaking R's usual rules for variable names; they're **non-syntactic** names.
To refer to these variables, you need to surround them with backticks, `` ` ``:
Você também pode notar que as colunas `ID Estudante` e `Nome Completo` estão cercadas por crases.
Isso ocorre porque elas contêm espaços, quebrando as regras usuais do R para nomes de variáveis; elas são nomes **não-sintáticos**.
Para se referir a essas variáveis, você precisa cercá-las com crases, `` ` ``:

```{r}
students |> 
estudantes |> 
  rename(
    student_id = `Student ID`,
    full_name = `Full Name`
    id_estudante = `ID Estudante`,
    nome_completo = `Nome Completo`
  )
```

An alternative approach is to use `janitor::clean_names()` to use some heuristics to turn them all into snake case at once[^data-import-1].

> [^data-import-1]: The [janitor](http://sfirke.github.io/janitor/) package is not part of the tidyverse, but it offers handy functions for data cleaning and works well within data pipelines that use `|>`.

Reviewer: Esse trecho também não está traduzido.

Reviewer 2: pelo que eu entendi, os trechos de nota de rodapé ficaram ao final do arquivo!

> Outra tarefa comum após a importação dos dados é a correção dos tipos de variáveis.

Reviewer: ```suggestion
Outra tarefa comum após a leitura dos dados é examinar os tipos de variáveis.
```

> Atualmente, `idade` é uma variável de caractere porque uma das observações é digitada como `cinco` em vez de um `5` numérico.

Reviewer: ```suggestion
Atualmente, `idade` é uma variável de caractere porque uma das observações está digitada como `cinco` em vez de um `5` numérico.
```

```

Uma nova função aqui é `if_else()`, que tem três argumentos.
O primeiro argumento `test` deve ser um vetor lógico.
O resultado conterá o valor do segundo argumento, `yes`, quando `test` é `TRUE`, e o valor do terceiro argumento, `no`, quando é `FALSE`.

> Aqui estamos dizendo que se `idade` é a *string* de caracteres `"cinco"`, faça-a `"5"`, e se não, mantenha-a como `idade`.

Reviewer: ```suggestion
Aqui estamos dizendo que se `idade` for o *string* de caracteres `"cinco"`, faça-o `"5"`, e se não, mantenha-o como `idade`.
```

)
```

Normalmente, `read_csv()` usa a primeira linha de dados como os nomes das colunas, o que é uma convenção muito comum.
Mas não é incomum que algumas linhas de metadados sejam incluídas no início do arquivo.

> Você pode usar `skip = n` para pular as primeiras `n` linhas ou usar `comment = "#"` para descartar todas as linhas que começam com `#`, por exemplo:

Reviewer: ```suggestion
Você pode usar `skip = n` para pular as primeiras `n` linhas ou usar `comment = "#"` para ignorar todas as linhas que começam com `#`, por exemplo:
```

```{r}
#| message: false

read_csv(
  "A primeira linha de metadados
  A segunda linha de metadados
  x,y,z
  1,2,3",
  skip = 2
)

read_csv(
  "# Um comentário que gostaria de ignorar
  x,y,z
  1,2,3",
  comment = "#"
)
```

Em outros casos, os dados podem não ter nomes de colunas.

> Você pode usar `col_names = FALSE` para dizer ao `read_csv()` para não tratar a primeira linha como cabeçalhos e, em vez disso, rotulá-los sequencialmente de `X1` a `Xn`:

Reviewer: ```suggestion
Você pode usar `col_names = FALSE` para dizer ao `read_csv()` para não tratar a primeira linha como cabeçalho e, em vez disso, rotular as colunas sequencialmente de `X1` a `Xn`:
```

)
```

> Alternativamente, você pode passar a `col_names` um vetor de caracteres que será usado como os nomes das colunas:

Reviewer: ```suggestion
Alternativamente, você pode fornecer à `col_names` um vetor de caracteres que será usado como os nomes das colunas:
```

)
```

Esses argumentos são tudo o que você precisa saber para ler a maioria dos arquivos CSV que você encontrará na prática.
(Para o restante, você precisará inspecionar cuidadosamente seu arquivo `.csv` e ler a documentação para os muitos outros argumentos de `read_csv()`.)

### Outros tipos de arquivos

Uma vez que você tenha aprendido a usar `read_csv()`, usar as outras funções do readr é simples; é apenas uma questão de saber qual função usar:

-   `read_csv2()` lê arquivos separados por ponto-e-vírgula.

>     Usa `;` ao invés de `,` para separar os campos e são comuns em países que usam `,` como marcador decimal.

Reviewer: ```suggestion
    Usa `;` ao invés de `,` para separar os campos, e são comuns em países que usam `,` como marcador decimal.
```

-   `read_tsv()` lê arquivos delimitados por tabulação.

> -   `read_delim()` lê arquivos com qualquer delimitador, tentando adivinhar automaticamente o delimitador se você não o especificar.

Reviewer: ```suggestion
-   `read_delim()` lê arquivos com qualquer delimitador, tentando supor automaticamente o delimitador se você não o especificar.
```

-   `read_fwf()` lê arquivos de largura fixa.
    Você pode especificar os campos por suas larguras com `fwf_widths()` ou por suas posições com `fwf_positions()`.

-   `read_table()` lê uma variação comum de arquivos de largura fixa onde as colunas são separadas por espaços em branco.

-   `read_log()` lê arquivos de log no estilo Apache.

### Exercícios

> 1.  Que função você usaria para ler um arquivo onde os campos são separados por `|`?

Reviewer: ```suggestion
1.  Que função você usaria para ler um arquivo onde os campos são separados por `\|`?
```

2.  Apart from `file`, `skip`, and `comment`, what other arguments do `read_csv()` and `read_tsv()` have in common?

2.  Além de `file`, `skip` e `comment`, quais outros argumentos `read_csv()` e `read_tsv()` têm em comum?

3.  What are the most important arguments to `read_fwf()`?

3.  Quais são os argumentos mais importantes para `read_fwf()`?

4.  Sometimes strings in a CSV file contain commas.
    To prevent them from causing problems, they need to be surrounded by a quoting character, like `"` or `'`. By default, `read_csv()` assumes that the quoting character will be `"`.
    To read the following text into a data frame, what argument to `read_csv()` do you need to specify?

> 4.  Às vezes, as *strings* em um arquivo CSV contêm vírgulas.

Reviewer: ```suggestion
4.  Às vezes, *strings* em um arquivo CSV contêm vírgulas.
```

>     Para evitar que elas causem problemas, elas precisam ser cercadas por um caractere de citação, como `"` ou `'`. Por padrão, `read_csv()` assume que o caractere de citação será `"`.

Reviewer: ```suggestion
    Para evitar que causem problemas, elas precisam ser acompanhadas por caracteres de citação, como `"` ou `'`. Por padrão, `read_csv()` assume que o carácter de citação será `"`.
```

Me parece que caracteres é o plural de carácter.

Reviewer 2: Dei uma pesquisada, e os dois estariam corretos! Mas em programação, é usado "caractere".

    read_csv("a;b\n1;3")
    ```

> 6.  Pratique se referir a nomes não-sintáticos no seguinte data frame por:

Reviewer: ```suggestion
6.  Pratique se referir a nomes não-sintáticos no seguinte data frame:
```

>     a.  Extrair a variável chamada `1`.

Reviewer: ```suggestion
    a.  Extraindo a variável chamada `1`.
```

>     b.  Plotar um gráfico de dispersão de `1` vs. `2`.

Reviewer: ```suggestion
    b.  Plotando um gráfico de dispersão de `1` vs. `2`.
```

>     c.  Criar uma nova coluna chamada `3`, que é `2` dividido por `1`.

Reviewer: ```suggestion
    c.  Criando uma nova coluna chamada `3`, que é `2` dividido por `1`.
```

>     d.  Renomear as colunas para `um`, `dois` e `três`.

Reviewer: ```suggestion
    d.  Renomeando as colunas para `um`, `dois` e `três`.
```

    ```{r}
    irritante <- tibble(
      `1` = 1:10,
      `2` = `1` * 2 + rnorm(length(`1`))
    )
    ```

## Controlando tipos de colunas {#sec-col-types}

> Um arquivo CSV não contém informações sobre o tipo de cada variável (ou seja, se é um lógico, número, *string*, etc.), então o readr tentará adivinhar o tipo.

Reviewer: ```suggestion
Um arquivo CSV não contém informações sobre o tipo de cada variável (ou seja, se é lógico, número, *string*, etc.), então readr tentará adivinhar o tipo.
```

Para cada coluna, ele pega os valores de 1.000[^data-import-3] linhas espaçadas uniformemente da primeira à última, ignorando os valores ausentes.
Ele então trabalha através das seguintes perguntas:

-   Contém apenas `F`, `T`, `FALSE` ou `TRUE` (ignorando a caixa alta)? Se sim, é um lógico.
-   Contém apenas números (por exemplo, `1`, `-4.5`, `5e6`, `Inf`)? Se sim, é um número.
-   Corresponde ao padrão ISO8601? Se sim, é uma data ou data-hora. (Voltaremos às data-horas em mais detalhes em @sec-creating-datetimes).
-   Caso contrário, deve ser uma *string*.

Você pode ver esse comportamento em ação neste exemplo simples:

```{r}
#| message: false

read_csv("
  logico,numerico,data,string
  TRUE,1,2021-01-15,abc
  false,4.5,2021-02-15,def
  T,Inf,2021-02-16,ghi
")
```

Essa heurística funciona bem se você tiver um conjunto de dados limpo, mas, na vida real, você encontrará uma seleção de falhas estranhas e belas.

### Valores ausentes, tipos de coluna e problemas

A forma mais comum de falha na detecção de coluna é quando uma coluna contém valores inesperados, e você obtém uma coluna de caracteres em vez de um tipo mais específico.

> Uma das causas mais comuns para isso é um valor ausente, registrado usando algo diferente do `NA` que o readr espera.

Reviewer: ```suggestion
Uma das causas mais comuns para isso é um valor ausente registrado usando algo diferente do `NA` que o readr espera.
```

)
```

Agora `read_csv()` relata que houve um problema e nos diz que podemos descobrir mais com `problems()`:

```{r}
problems(df)
```

Isso nos diz que houve um problema na linha 3, coluna 1, onde o readr esperava um número real, mas obteve um `.`.
Isso sugere que este conjunto de dados usa `.` para valores ausentes.
Então, definimos `na = "."`, a adivinhação automática tem sucesso, nos dando a coluna numérica que queremos:

```{r}
#| message: false

read_csv(simple_csv, na = ".")
```

### Tipos de colunas

readr fornece um total de nove tipos de coluna para você usar:

> -   `col_logical()` e `col_double()` leem lógicos e números reais. Eles são relativamente raramente necessários (exceto como acima), já que o readr geralmente os adivinhará para você.

Reviewer: ```suggestion
-   `col_logical()` e `col_double()` leem lógicos e números reais. É relativamente raro serem necessários (exceto como acima), já que o readr geralmente os adivinhará para você.
```

> -   `col_integer()` lê inteiros. Raramente distinguimos inteiros e duplos neste livro porque eles são funcionalmente equivalentes, mas ler inteiros explicitamente pode ocasionalmente ser útil porque ocupam metade da memória dos reais.

Reviewer 2: ```suggestion
-   `col_integer()` lê números inteiros (*integer*). Raramente distinguimos os números inteiros (*integer*) e reais (*doubles*) neste livro porque eles são funcionalmente equivalentes, mas ler números inteiros explicitamente pode ocasionalmente ser útil porque ocupam metade da memória dos números reais.
```

> -   `col_character()` lê *strings*. Isso pode ser útil para especificar explicitamente quando você tem uma coluna que é um identificador numérico, ou seja, longas séries de dígitos que identificam um objeto, mas não faz sentido aplicar operações matemáticas a ele. Exemplos incluem números de telefone, números de seguro social, números de cartão de crédito, etc.

Reviewer 2: ```suggestion
-   `col_character()` lê textos (*strings*). Isso pode ser útil para especificar explicitamente quando você tem uma coluna que é um identificador numérico, ou seja, longas séries de dígitos que identificam um objeto, mas não faz sentido aplicar operações matemáticas a ele. Exemplos incluem números de telefone, números de seguro social, números de cartão de crédito, etc.
```

> -   `col_character()` lê *strings*. Isso pode ser útil para especificar explicitamente quando você tem uma coluna que é um identificador numérico, ou seja, longas séries de dígitos que identificam um objeto, mas não faz sentido aplicar operações matemáticas a ele. Exemplos incluem números de telefone, números de seguro social, números de cartão de crédito, etc.

Reviewer: ```suggestion
-   `col_character()` lê *strings*. Isso pode ser útil para especificar explicitamente quando você tem uma coluna que é um identificador numérico, ou seja, longas séries de dígitos que identificam um objeto, mas onde não faz sentido aplicar operações matemáticas. Exemplos incluem números de telefone, números de seguro social, números de cartão de crédito, etc.
```

> -   `col_factor()`, `col_date()` e `col_datetime()` criam fatores, datas e data-horas, respectivamente; você aprenderá mais sobre esses tipos de dados quando chegarmos a eles em @sec-factors e @sec-dates-and-times.

Reviewer: Talvez _times_ necessite de uma tradução temporal mais genérica do que horas. Talvez **tempo**?

Reviewer 2: A forma usada é data-hora mesmo!

-   `col_number()` é um analisador numérico permissivo que ignorará componentes não numéricos, e é particularmente útil para moedas. Você aprenderá mais sobre ele em @sec-numbers.
-   `col_skip()` pula uma coluna para que ela não seja incluída no resultado, o que pode ser útil para acelerar a leitura dos dados se você tiver um arquivo CSV grande e quiser usar apenas algumas das colunas.

Também é possível substituir a coluna padrão trocando de `list()` para `cols()` e especificando `.default`:

```{r}
outro_csv <- "
x,y,z
1,2,3"

read_csv(
  outro_csv, 
  col_types = cols(.default = col_character())
)
```

> Outra função auxiliar é `cols_only()`, que lerá apenas as colunas que você especificar:

Reviewer: ```suggestion
Outra função auxiliar útil é `cols_only()`, que lerá apenas as colunas que você especificar:
```

```{r}
read_csv(
  outro_csv,
  col_types = cols_only(x = col_character())
)
```

## Importando dados de múltiplos arquivos {#sec-readr-directory}

> Às vezes seus dados estão divididos em vários arquivos em vez de estarem contidos em um único arquivo.

Reviewer: ```suggestion
Às vezes seus dados estão divididos em vários arquivos em vez de estarem contidos em um único.
```

Por exemplo, você pode ter dados de vendas para vários meses, com os dados de cada mês em um arquivo separado: `01-vendas.csv` para janeiro, `02-vendas.csv` para fevereiro e `03-vendas.csv` para março.

```{r}
#| message: false

> arquivos_vendas <- c("data/01-vendas.csv", "data/02-vendas.csv", "data/03-vendas.csv")

Reviewer: ```suggestion
arquivos_vendas <- c("dados/01-vendas.csv", "dados/02-vendas.csv", "dados/03-vendas.csv")
```

> read_csv(arquivos_vendas, id = "file")

Reviewer: ```suggestion
read_csv(arquivos_vendas, id = "arquivo")
```

```

> Mais uma vez, o código acima funcionará apenas se você tiver os arquivos CSV em uma pasta chamada `data` no seu projeto.

Reviewer: ```suggestion
Mais uma vez, o código acima funcionará apenas se você tiver os arquivos CSV em uma pasta chamada `dados` no seu projeto.
```

```{r}
#| eval: false

arquivos_vendas <- c(
  "https://raw.githubusercontent.com/cienciadedatos/pt-r4ds/traducao-pt-2ed/data/01-vendas.csv",
  "https://raw.githubusercontent.com/cienciadedatos/pt-r4ds/traducao-pt-2ed/data/02-vendas.csv",
  "https://raw.githubusercontent.com/cienciadedatos/pt-r4ds/traducao-pt-2ed/data/03-vendas.csv"
)
read_csv(arquivos_vendas, id = "arquivo")
```

O argumento `id` adiciona uma nova coluna chamada `arquivo` ao data frame resultante que identifica o arquivo de onde os dados vêm.
Isso é especialmente útil nas circunstâncias em que os arquivos que você está lendo não têm uma coluna de identificação que possa ajudá-lo a rastrear as observações até suas fontes originais.

Se você tiver muitos arquivos que deseja ler, pode ser complicado escrever seus nomes como uma lista.
Em vez disso, você pode usar a função `list.files()` para encontrar os arquivos para você, identificando um padrão nos nomes dos arquivos.
Você aprenderá mais sobre esses padrões em @sec-regular-expressions.

```{r}

> arquivos_vendas <- list.files("data", pattern = "vendas\\.csv$", full.names = TRUE)

Reviewer: ```suggestion
arquivos_vendas <- list.files("dados", pattern = "vendas\\.csv$", full.names = TRUE)
```

arquivos_vendas
```

## Exportando para um arquivo {#sec-writing-to-a-file}

> readr também vem com duas funções úteis para escrever dados de volta para o disco: `write_csv()` e `write_tsv()`.

Reviewer 2: ```suggestion
O pacote readr também vem com duas funções úteis para escrever dados de volta para o disco: `write_csv()` e `write_tsv()`.
```

Os argumentos mais importantes para essas funções são `x` (o data frame a ser salvo) e `file` (o local para salvá-lo).

> Você também pode especificar como os valores ausentes são escritos com `na`, e se você deseja acrescentar, via o argumento `append`, a um arquivo existente.

Reviewer 2: ```suggestion
Você também pode especificar como os valores ausentes são escritos com `na`, e se você deseja acrescentar as linhas em um arquivo existente, utilizando o argumento `append`.
```

```{r}
#| eval: false

write_csv(estudantes, "estudantes.csv")
```

> Agora, vamos ler esse arquivo CSV de volta.

Reviewer: ```suggestion
Agora, vamos ler esse arquivo CSV novamente.
```

Observe que as informações de tipo de variável que você acabou de configurar são perdidas quando você salva em CSV porque você está começando de novo com a leitura de um arquivo de texto simples:

```{r}
#| warning: false
#| message: false

estudantes
write_csv(estudantes, "estudantes-2.csv")
read_csv("estudantes-2.csv")
```

> 
> This makes CSVs a little unreliable for caching interim results---you need to recreate the column specification every time you load in.

Reviewer 2: ```suggestion

```

> There are two main alternatives:

Reviewer 2: ```suggestion

```

> Isso faz arquivos CSVs um pouco não confiáveis para armazenar resultados intermediários --- você precisa recriar a especificação das colunas toda vez que carrega os dados.

Reviewer 2: ```suggestion
Isso faz arquivos CSVs um pouco não confiáveis para armazenar resultados intermediários --- você precisa recriar a especificação das colunas toda vez que carrega os dados.
Existem duas formas principais:
```

1.  `write_rds()` e `read_rds()` são *wrappers* uniformes[^data-import-4] em torno das funções base `readRDS()` e `saveRDS()`.
    Eles armazenam dados no formato binário personalizado do R chamado RDS.

>     Isso significa que quando você recarrega o objeto, você está carregando o *mesmo* objeto R que você armazenou.

Reviewer: ```suggestion
    Isso significa que quando você recarrega o objeto, você está carregando *exatamente o mesmo* objeto R que você armazenou.
```

    ```{r}
    write_rds(estudantes, "estudantes.rds")
    read_rds("estudantes.rds")
    ```

2. O pacote arrow te permite importar e exportar arquivos parquet, um formato de arquivo binário rápido que pode ser compartilhado entre diferentes linguagens de programação.
    Voltaremos ao arrow com mais detalhes em @sec-arrow.

    ```{r}

>     #| eval: true

Reviewer 2: ```suggestion
    #| eval: false
```

    library(arrow)
    write_parquet(estudantes, "estudantes.parquet")

>     read_parquet("estudantes.parquet")

Reviewer 2: Lembrar de colocar o output aqui!

> `tribble()` is customized for data entry in code: column headings start with `~` and entries are separated by commas.

Reviewer 2: acho que isso não ficou traduzido

> `tribble()` is customized for data entry in code: column headings start with `~` and entries are separated by commas.

Reviewer 2: ```suggestion
`tribble()` é customizado para entrada de dados no código: os cabeçalhos das colunas começam com `~` e as entradas são separadas por vírgulas.
```

> This makes it possible to lay out small amounts of data in an easy to read form:

Reviewer 2: Acho que isso não ficou traduzido

> This makes it possible to lay out small amounts of data in an easy to read form:

Reviewer 2: ```suggestion
Isso torna possível dispor pequenas quantidades de dados em uma forma fácil de ler:
```

> Inserir os dados por coluna pode dificultar a visualização de como as linhas estão relacionadas, então uma alternativa é `tribble()`, abreviação de _**tr**ansposed t**ibble**_, que permite organizar seus dados linha por linha.

Reviewer 2: ```suggestion
Inserir os dados por coluna pode dificultar a visualização de como as linhas estão relacionadas, então uma alternativa é `tribble()`, abreviação de _**tr**ansposed t**ibble**_ (tibble transposta), que permite organizar seus dados linha por linha.
```

)
```

## Resumo

> Neste capítulo, você aprendeu a como carregar arquivos CSV com `read_csv()` e a como inserir seus próprios dados com `tibble()` e `tribble()`.

Reviewer: ```suggestion
Neste capítulo, você aprendeu como carregar arquivos CSV com `read_csv()` e como inserir seus próprios dados com `tibble()` e `tribble()`.
```

Você aprendeu como arquivos CSV funcionam, alguns dos problemas que você pode encontrar e como superá-los.

> Voltaremos à importação de dados algumas vezes neste livro: @sec-import-spreadsheets do Excel e Google Planilhas, @sec-import-databases mostrará como carregar dados de bancos de dados, @sec-arrow de arquivos parquet, @sec-rectangling de JSON e @sec-scraping de sites.

Reviewer: ```suggestion
Voltaremos à importação de dados algumas vezes neste livro: @sec-import-spreadsheets do Excel e do Google Planilhas, @sec-import-databases mostrará para você como carregar dados de bancos de dados, @sec-arrow de arquivos parquet, @sec-rectangling de JSON e @sec-scraping de sites.
```

Estamos quase no final desta seção do livro, mas há um último tópico importante a ser coberto: como obter ajuda.

> Então, no próximo capítulo, você aprenderá alguns bons lugares para procurar ajuda, como criar um reprex[^data-import-5] para maximizar suas chances de obter boa ajuda e alguns conselhos gerais sobre como acompanhar o mundo do R.

Reviewer 2: ```suggestion
Então, no próximo capítulo, você aprenderá alguns bons lugares para procurar ajuda, como criar um exemplo reprodutível (*reprex*)[^data-import-5] para maximizar suas chances de obter boa ajuda e alguns conselhos gerais sobre como acompanhar o mundo do R.
```
